tensorflow/core/grappler/optimizers/convert_to_bfloat16.cc

```diff
@@ -0,0 +1,167 @@
+        (GetNodeAttr(n->def(), "T", &T) == Status::OK()) & (T == DT_FLOAT) &&
+        CanOpSupportBFloat16(n)) {
+      // Rewrite this op if we want to always rewrite it or if all the inputs of
+      // this op are in BFloat16 type.
+      string action = "";
+      if (RewriteIfAllInputsInBFloat16Op(n)) {
+        if (AreAllInputsInBFloat16(n)) {
+          action = "changing type to BFLOAT16";
+          ChangeDataType(n, DT_BFLOAT16);
+        } else {
+          action = "skipping BFLOAT16 type conversion";
+        }
+      } else if (AlwaysRewriteOp(n)) {
+        action = "changing type to BFLOAT16";
+        ChangeDataType(n, DT_BFLOAT16);
+      }
+      VLOG(1) << name() << ": " << action << ": op=" << n->name()
+              << ", op_type=" << n->type_string();
+      ;
+      if (do_logging_) {
+        CONVERTER_LOG() << name() << ": " << action << ": op=" << n->name()
+                        << ", op_type=" << n->type_string() << std::endl;
+      }
+    }
+
+    // Traverse all input edges and insert cast nodes on an edge if its
+    // source type is different than destination type.
+    for (const Edge* e : n->in_edges()) {
+      // No need to do anything for coontrol edges since current node will not
+      // be deleted anyway.
+      if (e->IsControlEdge()) continue;
+      DataType src_dtype = e->src()->output_type(e->src_output());
+      DataType dst_dtype = e->dst()->input_type(e->dst_input());
+      // Only insert Cast nodes between float32 and bfloat16 because we dont
+      // anticipate bfloat16 usage in non-FP32 models.
+      if (src_dtype != dst_dtype &&
+          (src_dtype == DT_FLOAT || src_dtype == DT_BFLOAT16) &&
+          (dst_dtype == DT_FLOAT || dst_dtype == DT_BFLOAT16)) {
+        string cast_type = "";
+        if (src_dtype == DT_FLOAT)
+          cast_type = "FP32toBF16";
+        else
+          cast_type = "BF16toFP32";
+        VLOG(1) << name() << ": inserting " << cast_type << " Cast on edge "
+                << e->src()->name() << ":" << e->src_output() << " and "
+                << e->dst()->name() << ":" << e->dst_input();
+        if (do_logging_) {
+          CONVERTER_LOG() << name() << ": inserting " << cast_type
+                          << " Cast on edge " << e->src()->name() << ":"
+                          << e->src_output() << " and " << e->dst()->name()
+                          << ":" << e->dst_input() << std::endl;
+        }
+
+        TF_RETURN_IF_ERROR(InsertCastNode(g, e, src_dtype, dst_dtype));
+      }
+    }
+  }
+  return Status::OK();
+}
+
+Status BFloat16Converter::Optimize(Cluster* cluster, const GrapplerItem& item,
+                                   GraphDef* output_graphdef) {
+  if (do_logging_) {
+    CONVERTER_LOG() << name() << ": is enabled" << std::endl;
+  }
+
+  // TODO(nhasabni): what to do with other args?
+  Graph output_graph(OpRegistry::Global());
+  Status status = ConvertGraphDefToGraph(GraphConstructorOptions(), item.graph,
```

Comment:
The round-trip via Graph adds quite a bit of overhead here, but we can probably tolerate it. Grappler passes are generally written directly in terms of manipulations of the GraphDef, using e.g. the helper class MutableGraphView here:

https://github.com/tensorflow/tensorflow/blob/01353dbe4b11737d47a0b25db0aedd725dfac241/tensorflow/core/grappler/utils/graph_view.h#L380

However, since you already went this route, I don't insist.

Reply:
I've marked this as TODO. I will see if I can address this in the next round over this file.